### ISSUE TITLE
tests/weak-ephe-final/finaliser_handover is flaky 

### ISSUE THREAD

**Issue by gasche** [ocaml/ocaml#12345]
This tests seems to fail from time to time on the github CI, for different reasons and on different systems. The test could be wrong, but it may also catch one or several bugs. In any case, those occasional failure are costing developer time -- it is tedious to lookup CI failures, and brain-consuming to make the guess that they are unrelated to the current PR.

- for #12222, #12308, #12338, #12381, #12507 the test failed on a linux machine (both bytecode and nativee failures where observed), with the error

    > [02] file runtime/domain.c; line 1847 ### Assertion failed: caml_gc_phase == Phase_sweep_and_mark_main

- for #12344 the test failed for native code on a macos machine, with the error

   > Fatal error: exception File "finaliser_handover.ml", line 39, characters 2-8: Assertion failed

- for #12551 and #12615 the test failed in bytecode on a Windows machine

**Comment by kayceesrk** [comment]
I've assigned this issue to myself. The first assertion failure either indicates a bug or an incorrect assertion. I'll spend some time to investigate this.

**Comment by xavierleroy** [comment]
Why don't we disable this test until it is analyzed?  A CI that fails randomly is useless.

**Comment by gasche** [comment]
If @kayceesrk happens to not have the time to look at this, maybe we could consider asking @fabbing or @NickBarnes for help?

@xavierleroy why not indeed; my question above is in the mindset of getting the bug fixed, rather than just denoising the CI.

**Comment by gasche** [comment]
Today @Octachron and @damiendoligez and myself tried to look at this. We timed out for now, but here is what we understand of the issue.

The bug comes from the following code in domain.c:

```c

static void handover_finalisers(caml_domain_state* domain_state)
{
  struct caml_final_info* f = domain_state->final_info;
  if (f->todo_head != NULL || f->first.size != 0 || f->last.size != 0) {
    /* have some final structures */
    if (caml_gc_phase != Phase_sweep_and_mark_main) {
      /* Force a major GC cycle to simplify constraints for
       * handing over finalisers. */
      caml_finish_major_cycle();
      CAMLassert(caml_gc_phase == Phase_sweep_and_mark_main);
    }
    caml_add_orphaned_finalisers (f);
    /* Create a dummy final info */
    domain_state->final_info = caml_alloc_final_info();
  }
  caml_final_domain_terminate(domain_state);
}
```

The "handover" function is responsible, on domain termination, for adding the finalizers of the current (terminating) domain to a global list of "orphaned" finalizers. 

The logic in this code is as follows:
- we only want to perform the handover if we are in a specific GC phase: `sweep_and_mark_main` (why? the code says "to simplify constraints", but what does this mean? documentation here would be welcome)
- if we are not in this phase, force to finish the current major GC to get back in this phase (`caml_finish_major_cycle();`)

The assertion that we are then in the desired phase fails.

What we believe happens is that finishing the current major GC cycle puts us in the correct phase, but some other domain makes progress in parallel and, by the time we get to the assert (or to the orphaning outside debug mode) we are not in this phase anymore.

It is not obvious how to fix this bug. An idea would be to delegate the orphaning work to the `finish_major_cycle` code: domains can put themselves in the state "I need to handover my finalizers", and the major-GC logic could call `add_orphaned_finalizerrs` at the right time, when we are still in the stop-the-world phase and the major GC cannot make concurrent progress.

One issue with this proposal is that it requires forcing a major GC cycle at each domain termination: we cannot rely on the `if (caml_gc_phase != Phase_sweep_and_mark_main)` test (as we believe it is not robust to concurrent GC work), we have to call `finish_major_cycle` unconditionally.

It would be nice to better understand the constraints on finalizers / when the handover can happen. What are the reasons why we want to be in a specific GC phase? cc @kayceesrk 

(I thought at first that the constraint tries to avoid a race on the global list of orphaned finalizers, but that list seems to be protected by a mutex anyway.)

**Comment by gasche** [comment]
I went back to this issue today and indirectly started #12591 -- which touches the same part of the codebase but fixes another issue, a data race on `orph_structs`.

**Comment by gasche** [comment]
@kayceesrk this test (finaliser_handover) is failing on a somewhat-regular basis due to a design issue in the multicore runtime's handling of finalisers that we have diagnosed/identified in https://github.com/ocaml/ocaml/issues/12345#issuecomment-1651865960 -- the code of `handover_finalisers` is wrong. It would be nice to fix the runtime (and thus the test).

I would be willing to help in implementing a fix, but I suspect that designing a good fix requires understanding the invariants that are being assumed on the state of finalisers when orphaning them, and those assumptions are not documented in the runtime code that I could find. Do you know who would be available to explain/document this, or propose a fix for the issue themselves?

**Comment by gasche** [comment]
Two other recent failures: #12692 and #12703.

**Comment by kayceesrk** [comment]
I'm trying to reproduce this failure on my local machine. I'll need a bit more time on this.

**Comment by Octachron** [comment]
In my experience, the failure can be consistently reproduced by launching a hundred of tests in parallel. I can share my reproduction script by next week if you wish.

**Comment by kayceesrk** [comment]
That's useful to know. Happy to get hold of your script. 

Perhaps even commit the script under `tools/`? We already have a few tools such as `gdb-macros` and `gdb_ocamlrun.py`.

**Comment by gasche** [comment]
There is a new failure in #12707 (on the 'i386' runner) which I had not seen before:

```
... testing 'finaliser_handover.ml' with 2 (bytecode) Timeout expired, killing all child processes
Process 40549 got signal 9(Killed), no core dumped
=> failed (Running program /__w/ocaml/ocaml/testsuite/tests/weak-ephe-final/_ocamltest/tests/weak-ephe-final/finaliser_handover/ocamlc.byte/finaliser_handover.byte without any argument: command
/__w/ocaml/ocaml/testsuite/tests/weak-ephe-final/_ocamltest/tests/weak-ephe-final/finaliser_handover/ocamlc.byte/finaliser_handover.byte 
failed with exit code -9)
```

I have the vague impression is that the failure is happening more often this week than it was in the past, possibly due to an unrelated runtime change that makes the bad-scheduling scenarios more likely. (In the present case I think we have a good understanding of what the issue is, so new data is not necessarily that helpful.)

**Comment by gasche** [comment]
In #12708 I propose to disable the finaliser_handover test from the CI for now.

1. It fails often enough to be a nuisance.
2. I believe that we understand the cause of the failure.
3. The fix may take some time.
4. I believe that we are not going to forget about this bug entirely just because we are not notified by the CI anymore.

**Comment by gasche** [comment]
Thanks to KC's #12710, the known issue with the test has been fixed, so closing here.
